ams/optimizer/spark-optimizer/src/main/java/com/netease/arctic/optimizer/spark/SparkExecutor.java

```diff
@@ -0,0 +1,128 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.optimizer.spark;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.netease.arctic.ams.api.OptimizingTask;
+import com.netease.arctic.ams.api.OptimizingTaskResult;
+import com.netease.arctic.optimizer.common.AbstractOptimizerOperator;
+import com.netease.arctic.optimizing.RewriteFilesInput;
+import com.netease.arctic.optimizing.TableOptimizing;
+import com.netease.arctic.utils.SerializationUtil;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class SparkExecutor extends AbstractOptimizerOperator {
+  private static final Logger LOG = LoggerFactory.getLogger(SparkExecutor.class);
+  private final ThreadPoolExecutor executorService;
+  private final SparkFunction sparkFunction;
+  private final JavaSparkContext jsc;
+  private final int threadId;
+
+  public SparkExecutor(JavaSparkContext jsc, SparkOptimizerConfig config, int threadId) {
+    super(config);
+    this.jsc = jsc;
+    this.threadId = threadId;
+    this.sparkFunction = new SparkFunction(config, threadId);
+    ThreadFactory executorFactory =
+        new ThreadFactoryBuilder()
+            .setDaemon(false)
+            .setNameFormat("spark-optimizing-executor-%d")
+            .build();
+    this.executorService =
+        new ThreadPoolExecutor(
+            0,
+            config.getExecutionParallel(),
+            30L,
+            TimeUnit.SECONDS,
+            new ArrayBlockingQueue<>(config.getExecutionParallel()),
+            executorFactory);
+  }
+
+  public ThreadPoolExecutor getExecutorService() {
+    return this.executorService;
+  }
+
+  public void submitOptimizingTask(OptimizingTask task) {
+    executorService.execute(
+        () -> {
+          try {
+            String threadName = Thread.currentThread().getName();
+            long startTime = System.currentTimeMillis();
+            LOG.info("Now [{}] execute task {}", threadName, task);
+            ImmutableList<OptimizingTask> of = ImmutableList.of(task);
+            String jobDesc =
+                String.format("%s | TableName(%s)", task.getTaskId(), parseFullTableName(task));
+            jsc.setJobDescription(jobDesc);
+            List<OptimizingTaskResult> result = jsc.parallelize(of, 1).map(sparkFunction).collect();
+            result.forEach(SparkExecutor.this::completeTask);
```

Comment:
Why perform optimization tasks via thread pool? Seems like a simple matter of wrapping the execution into Spark rather than utilizing Spark's distributed processing capabilities?

Reply:
Yes, we just reuse the impl of OptimizingExecutor#execute in core module. But by setting spark.dynamicAllocation.enabled and parallel of the optimizer, the compute resource can be fully used and is more flexible.

Reply:
Okay, so SparkExecutor just submit the tasks to spark application. That's make sense